### PR TITLE
GEODE-8537: Solve memory increase when LRU eviction is enabled

### PR DESCRIPTION
**Pull request by gaussianrecurrence**
- Whenever LRU eviction was enabled it was noted an slight increase in the memory usage.
  Specifically in an scenario in which a set of entries are continously created and destroyed.
- Problem was that entries within LRUList where inserted but not removed until LRU
   eviction happened and in the described case above, that was never.
- Solution was to replace the LRUList by a refactored version called LRUQueue and
  also to remove the entries from the queue upon destroy or invalidation.
- Also a dead-lock between EvictionController and EvictionThread has been solved
  resulting in a complete refactor.
- Unit tests have been added for the LRUQueue.
- Removed old integration TS for LRUList as it has now been replace by an UT TS.
- Integration test have been added in the new integration test for the LRU eviction.
- Also a wrongly implemented LRU eviction test was removed from the old integration tests.

### CODE REVIEW

**Comment by gaussianrecurrence** [comment]
Maybe @moleske @pivotal-jbarrett @mreddington @pdxcodemonkey you could throw some comments?

Thanks!

**Review by pdxcodemonkey — CHANGES_REQUESTED** [review]
Excellent-looking change, love all the new tests, thanks!  Just the one small thing about naming, and go with someone else's queue implementation if we can.

File: cppcache/src/LRUQueue.hpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+namespace geode {
+namespace client {
+
+class MapEntryImpl;
+
+/**
+ * This class holds a queue of entries sorted by its use order
+ * @note All accesses to the queue are mutually exclusive
+ */
+class LRUQueue {
```
[the remark this answers is not in the document]
Comment (gaussianrecurrence) [reply cppcache/src/LRUQueue.hpp:40]: Regarding why it is needed a custom written queue. There are several reason why:

1. I wanted to encapsulate LRUProperties iterator management inside a class, so any code using the LRUQueue does not have to take care of that.
2. Usually, queues does not allow to remove entries at a random position, which would not make possible to clear deleted entries, which is the whole point of this PR.
3. There is no queue implementation that allows to take an entry and move it to its end.

What I am not sure about is wether to use a spinlock, an standard mutex or put some effort and try to write a completly lock-free thread-safe LRU queue :S

[the remark this answers is not in the document]
Comment (pdxcodemonkey) [reply cppcache/src/LRUQueue.hpp:40]: I'd just use a standard mutex.  Thanks for the clarification on the need for the LRUQueue, I hadn't paid close enough attention to what it was doing and missed the random-access feature.

**Review by pdxcodemonkey — APPROVED** [review]

**Pull request merged**